url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/Utils/ConsoleWriter.java

```diff
@@ -0,0 +1,226 @@
+				}
+			}
+		}
+
+		private void stopConsoleWriter() {
+			goOn = false;
+			this.interrupt();
+		}
+	}
+
+	/**
+	 * Start consolewriter scheduler thread. Should be called once at init time.
+	 */
+	public synchronized void startThread() {
+		print("Thread started.");
+		if (consoleWriterThread == null) {
+			consoleWriterThread = new ConsoleWriterThread("SERVER");
+			consoleWriterThread.start();
+		} else {
+			consoleWriterThread.interrupt();
+		}
+	}
+
+	/** Stop consolewriter scheduler thread. */
+	public void stopThread() {
+		print("Thread stopped");
+		if (consoleWriterThread != null) {
+			consoleWriterThread.stopConsoleWriter();
+			consoleWriterThread = null;
+		}
+	}
+
+	/**
+	 * Asks the server to identify this console. Puts the console code (generated by server) to a static variable to be
+	 * used among all classes.
+	 * 
+	 * A Console only needs to be started **once** per console!
+	 */
+	public void startConsole() {
+		if (consoleID.equals("")) {
+			consoleID = sendData("start=y");
+		}
+		System.out.println("===========================");
+		System.out.println("Console started: " + consoleID);
+		System.out.println("===========================");
+	}
+
+	/*
+	 * Messages
+	 */
+
+	/**
+	 * Called by agents in their print methods. Wraps the message around with source, and send with sendData()
+	 * 
+	 * @param sourceName
+	 *            agent's name
+	 * @param message
+	 *            the message to be sent
+	 */
+	public void sendMessage(String sourceName, String message) {
+		message = "[" + sourceName + "] " + message;
+		if (consoleID != "") {
+			message += "&consoleID=" + consoleID;
+		}
+		// System.out.println("Sending " + message);
+		messageQueue.add("message=" + message);
+		stateChanged();
+	}
+
+	/**
+	 * Alternatively, you can specify the source name by specifying in constructor, and pass only the message as param
+	 * when sending message.
+	 */
+	public void sendMessage(String message) {
+		if (name != null) {
+			message = "[" + name + "] " + message;
+		}
+		if (consoleID != "") {
+			message += "&consoleID=" + consoleID;
+		}
+		// System.out.println("Sending " + message);
+		messageQueue.add("message=" + message);
+		stateChanged();
+	}
+
+	/*
+	 * Scheduler
+	 */
+	public boolean sendingMessage() {
+		if (messageQueue.size() > 0) {
+			sendData(messageQueue.remove());
+			return true;
+		}
+
+		return false;
+	}
+
+	/*
+	 * Actions
+	 */
+
+	/**
+	 * Establishes a connection to the server and sends data.
+	 * 
+	 * @param data
+	 *            params to be sent
+	 * @return response from the sever
+	 */
+	public String sendData(String data) {
+		
+		try {
+			URLConnection connection = new URL("http://ptz-debug.appspot.com/listen/").openConnection();
+			connection.setDoOutput(true);
+			connection.setRequestProperty("Accept-Charset", "UTF-8");
+			connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded;charset=UTF-8");
+
+			OutputStream output = connection.getOutputStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *output*. Currently, there are execution paths that do not contain closure statements, for example, when OutputStream.write throws an exception. Ensure *output* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).